lib/Sema/CSSolver.cpp

```diff
@@ -652,6 +652,18 @@ bool ConstraintSystem::tryTypeVariableBindings(
         type = type->reconstituteSugar(/*recursive=*/false);
       }
 
+      // Instead of binding collection types directly let's try to
+      // bind using temporary type variables substituted for element
+      // types, that's going to ensure that subtype relationship is
+      // always preserved.
+      if (isConcreteCollectionType(type)) {
+        auto *BGT = type->getAs<BoundGenericType>();
```

Comment:
FWIW, in cases where `getAs` should never fail, you should use `castTo` to always get a non-null pointer (or assert if it's not of the right type).

Reply:
Thanks, I didn't know that!